pom.xml

```diff
@@ -2476,7 +2476,9 @@
                 <configuration>
                     <properties>
                         <configurationParameters>junit.jupiter.execution.timeout.thread.mode.default = SEPARATE_THREAD
-                            junit.jupiter.extensions.autodetection.enabled = true</configurationParameters>
+                            junit.jupiter.extensions.autodetection.enabled = true
+                            junit.jupiter.execution.parallel.enabled = true
+                            junit.jupiter.execution.parallel.mode.default = concurrent</configurationParameters>
```

Comment:
Add:
```
junit.jupiter.execution.parallel.config.strategy=fixed
junit.jupiter.execution.parallel.config.fixed.parallelism=${air.test.thread-count}
```

Reply:
I'll try that if there are issues, but the dynamic strategy seems like a better choice, as it can take advantage of more parallelism if the tests run on a machine with more cores.

Reply:
For some modules we've reduced parallelism

Reply:
I only see `air.test.thread-count` set in the root POM. Several modules set `air.test.parallel` to `instances` in order to workaround a TestNG bug.

Reply:
> I only see `air.test.thread-count` set in the root POM.

There are 4 places where we set `<threadCount>` directly on surefire plugin configuration.
(actually to increase parallelism)

```
$ git grep '<threadCount>'
plugin/trino-redshift/pom.xml:                            <threadCount>4</threadCount>
testing/trino-faulttolerant-tests/pom.xml:                            <threadCount>4</threadCount>
testing/trino-faulttolerant-tests/pom.xml:                            <threadCount>4</threadCount>
testing/trino-faulttolerant-tests/pom.xml:                            <threadCount>4</threadCount>
```

Reply:
> Several modules set air.test.parallel to instances in order to workaround a TestNG bug.

BTW what are the JUnit ordering guarantees for instances when executing tests concurrently?
I couldn't find any concrete information on this, but would be good to know how many concurrent tests classes can be running, especially for resource hungry tests (eg spinning up docker containers)

> but the dynamic strategy seems like a better choice, as it can take advantage of more parallelism if the tests run on a machine with more cores.

That's true. More precisely, it should do more parallelism when there is more cores **and** more memory.

Reply:
There are no guarantees, but you can use resource locks to limit concurrency.

Reply:
> There are no guarantees, but you can use resource locks to limit concurrency.

When running tests locally on a laptop with 16 cores, how can i limit the memory used by the test process?
Will `-Xmx8192m` set in `.mvn/jvm.config` suffice in such case?


> There are 4 places where we set `<threadCount>` directly on surefire plugin configuration.
> (actually to increase parallelism)

All these cases are apparently fault-tolerant tests and these are still on TestNG, so this may not be a problem today.
They use increased parallelism (to 4) because test duration is not CPU-bound. Some tests are simply latency bound (eg fault-tolerant, Redshift's or BigQuery's)
AFAIC, github runners use 2 cores, so with the dynamic strategy the tests may be 2x slower, once migrated to JUnit.